src/Middleware/HttpLogging/src/W3CLogger.cs

```diff
@@ -0,0 +1,177 @@
+                        var elapsed = DateTime.Now.Subtract(dt);
+                        elements[BitOperations.Log2((int)W3CLoggingFields.TimeTaken)] = elapsed.TotalMilliseconds.ToString(CultureInfo.InvariantCulture);
+                        break;
+                    case nameof(HeaderNames.Server):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ServerName)] = val.Trim();
+                        break;
+                    case nameof(ConnectionInfo.RemoteIpAddress):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ClientIpAddress)] = val.Trim();
+                        break;
+                    case nameof(ConnectionInfo.LocalIpAddress):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ServerIpAddress)] = val.Trim();
+                        break;
+                    case nameof(ConnectionInfo.LocalPort):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ServerPort)] = val.Trim();
+                        break;
+                    case nameof(HttpContext.User):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.UserName)] = val.Trim();
+                        break;
+                    case nameof(HeaderNames.Cookie):
+                        // Cookie can have whitespace - we replace whitespace characters with the '+' character
+                        elements[BitOperations.Log2((int)W3CLoggingFields.Cookie)] = replaceWhitespace(val.Trim());
+                        break;
+                    default:
+                        break;
+                }
+            }
+            var sb = new ValueStringBuilder();
+            var firstElement = true;
+            for (var i = 0; i < elements.Length; i++)
+            {
+                if (_options.CurrentValue.LoggingFields.HasFlag((W3CLoggingFields)(1 << i)))
+                {
+                    if (!firstElement)
+                    {
+                        sb.Append(' ');
+                    }
+                    else
+                    {
+                        firstElement = false;
+                    }
+                    // If the element was not logged, or was the empty string, we log it as a dash
+                    if (string.IsNullOrEmpty(elements[i]))
+                    {
+                        sb.Append('-');
+                    }
+                    else
+                    {
+                        sb.Append(elements[i]);
+                    }
+                }
+            }
+            return new LogMessage(DateTimeOffset.Now, sb.ToString());
+        }
+
+        // Modified from https://www.codeproject.com/Articles/1014073/Fastest-method-to-remove-all-whitespace-from-Strin
+        private string replaceWhitespace(string entry)
+        {
+            var len = entry.Length;
+            var src = entry.ToCharArray();
```

Comment:
This is always allocating a new char array and string, even if no changes are required.

What I'd do is make ToCharArray happen at the point the first replacement is required.

Then at the end, you'd only create a new string from the char array if it isn't null. Otherwise, return the original string.